Add GPUSupportEnabled to ECS configurations

arnaldo2792 opened the pull request:
**Issue number:**
Closes #2086 


**Description of changes:**

In preparation for #1074, this adds the `GPUSupportEnabled` flag as `false` for the existing ECS variant.

```
Now the value of the `GPUSupportEnabled` flag will be set at
compile-time depending on the variant being built.
```


**Testing done:**

I checked that the `ecs.config.json` file was created with the expected flag set as `false` for the general purpose variant ECS variant:

```
[root@admin]# cat /.bottlerocket/rootfs/etc/ecs/ecs.config.json  | jq
{
  "Cluster": "bottlerocket",
  "InstanceAttributes": {
    "bottlerocket.variant": "aws-ecs-1"
  },
  "PrivilegedDisabled": true,
  "AvailableLoggingDrivers": [
    "json-file",
    "awslogs",
    "none"
  ],
  "TaskIAMRoleEnabled": true,
  "TaskIAMRoleEnabledForNetworkHost": true,
  "SELinuxCapable": true,
  "OverrideAWSLogsExecutionRole": true,
  "TaskENIEnabled": true,
  "GPUSupportEnabled": false
}

```


**Terms of contribution:**

By submitting this pull request, I agree that this contribution is dual-licensed under the terms of both the Apache License, version 2.0, and the MIT license.

arnaldo2792 committed:
ecs-settings-applier: add GPUSupportEnabled to ECS configurations

Now the value of the `GPUSupportEnabled` flag will be set at
compile-time depending on the variant being built.

Signed-off-by: Arnaldo Garcia Rincon <agarrcia@amazon.com>

etungsten commented on sources/api/ecs-settings-applier/src/ecs.rs:
```diff
@@ -101,6 +109,8 @@ async fn run() -> Result<()> {
         // awsvpc mode is always available
         task_eni_enabled: true,
+
+        gpu_support_enabled: GPU_SUPPORT_ENABLED,
```
nit: I wonder if we can make the condition more clear by using the [`cfg!`](https://doc.rust-lang.org/std/macro.cfg.html) macro here:
```suggestion
        gpu_support_enabled: !(cfg!(variant_type = "general_purpose")),
```

then we can ditch the intermediary const bool.

arnaldo2792 replied:
I can give this a spin :+1:

etungsten commented on sources/api/ecs-settings-applier/src/ecs.rs:
```diff
@@ -64,6 +71,7 @@ pub(crate) async fn main() -> () {
 
 async fn run() -> Result<()> {
     let args = parse_args(env::args());
+    SimpleLogger::init(LevelFilter::Info, LogConfig::default()).context(error::LoggerSnafu)?;
```
I guess logging was never working prior to this change?

nit: Can we move these changes to a separate commit?

arnaldo2792 replied:
That's correct, nothing was logged out. I can put this in a different commit

arnaldo2792 committed:
ecs-settings-applier: fix logging

Nothing was being logged by `ecs-settings-applier` since it didn't
configure the program's default logger.

Signed-off-by: Arnaldo Garcia Rincon <agarrcia@amazon.com>

arnaldo2792 commented:
Last two forced pushes address:

- Remove constant value and use `!cfg` instead
- Separate logging change into its own commit

arnaldo2792 commented:
(Forced push to rebase develop)

webern commented on sources/api/ecs-settings-applier/build.rs:
```diff
@@ -9,6 +9,17 @@ fn main() {
     println!("cargo:rerun-if-env-changed=VARIANT");
     if let Ok(variant) = env::var("VARIANT") {
         println!("cargo:rustc-cfg=variant=\"{}\"", variant);
+        let parts = variant.split('-').collect::<Vec<&str>>();
+        println!(
+            "cargo:rustc-cfg=variant_family=\"{}\"",
+            parts[0..2].join("-")
+        );
+        let variant_type = if parts.len() > 3 {
+            parts[3]
+        } else {
+            "general_purpose"
+        };
+        println!("cargo:rustc-cfg=variant_type=\"{}\"", variant_type);
```
It doesn't have to be part of this PR, but we should generalize this logic. Possibly buildsys could have a lib.rs that we use for things like this.

bcressey commented on sources/api/ecs-settings-applier/build.rs:
```diff
@@ -9,6 +9,17 @@ fn main() {
+            "cargo:rustc-cfg=variant_family=\"{}\"",
+            parts[0..2].join("-")
+        );
+        let variant_type = if parts.len() > 3 {
+            parts[3]
+        } else {
+            "general_purpose"
+        };
```
Why not check for NVIDIA specifically? Is it because we'd also set `GPUSupportEnabled` for AMD and Inferentia?

arnaldo2792 replied:
I didn't see support for AMD in the ECS agent, but to enable inferentia support, a different flag has to be (`InferentiaSupportEnabled`) set and we will set it based on the variant's name.

bcressey replied:
It might be better to pin it to NVIDIA if `GPUSupportEnabled` is synonymous with that hardware.

arnaldo2792 replied:
But we still need to know what "variant type" is built, since we could have `aws-k8s-*-inferentia` so that in `ecs-settings-applier` we do:

```rust
{
    GPUSupportEnabled : cfg!(variant_type = "nvidia"), 
    InferentiaSupportEnabled: cfg!(variant_type = "inferentia")
}
```

arnaldo2792 commented:
Forced push changes the condition to set `GPUSupportEnabled`
